bundle/edu.gemini.shared.util/src/main/java/edu/gemini/shared/util/immutable/ImEither.java

```diff
@@ -1,120 +1,47 @@
-    }
-
-    // TODO: an ImEither(None, None) wouldn't make any sense.  Reimplement
-    // TODO: with an ImEither interface and Left and Right implementations.
-
-    private ImEither(final Option<L> left, final Option<R> right) {
-        this.left  = left;
-        this.right = right;
-    }
-
-    /**
-     * Folds
-     */
-    public <T> T fold(final Function1<? super L, ? extends T> leftFunc,
-                      final Function1<? super R, ? extends T> rightFunc) {
-
-        return left.map(leftFunc).getOrElse(() -> right.map(rightFunc).getValue());
-    }
-
-    public <T> T foldRight(final T zero, final Function2<? super R, ? super T, ? extends T> rightFunc) {
-        final Option<T> result = right.map(r -> rightFunc.apply(r, zero));
-        return result.getOrElse(zero);
-    }
-
-    /**
-     * Existential qualifiers
-     */
-    public boolean exists(final PredicateOp<? super R> rightPred) {
-        return right.exists(rightPred);
-    }
-
-    public boolean forall(final PredicateOp<? super R> rightPred) {
-        return right.forall(rightPred);
-    }
-
-    /**
-     * foreach
-     */
-    public void foreach(final ApplyOp<? super R> rightConsumer) {
-        right.foreach(rightConsumer);
-    }
-
-    public void biForeach(final ApplyOp<? super L> leftConsumer,
-                          final ApplyOp<? super R> rightConsumer) {
-        left.foreach(leftConsumer);
-        right.foreach(rightConsumer);
-    }
-
-    /**
-     * Maps
-     */
-    public <S,T> ImEither<S,T> biMap(final Function1<? super L, ? extends S> leftMap,
-                                     final Function1<? super R, ? extends T> rightMap) {
-        return new ImEither<>(left.map(leftMap), right.map(rightMap));
-    }
-
-    public <T> ImEither<T,R> leftMap(final Function1<? super L, ? extends T> leftFunc) {
-        return new ImEither<>(left.map(leftFunc), right);
-    }
-
-    public <T> ImEither<L,T> map(final Function1<? super R, ? extends T> rightFunc) {
-        return new ImEither<>(left, right.map(rightFunc));
-    }
-
-    public <T> ImEither<L,T> flatMap(final Function1<? super R, ImEither<L, T>> rightFunc) {
-        return left.isDefined() ? ImEither.<L,T>left(left.getValue()) : rightFunc.apply(right.getValue());
-    }
-
-    /**
-     * Accessors and other assorted functions
-     */
-    public Option<L> getLeft() {
-        return left;
-    }
-
-    public Option<R> getRight() {
-        return right;
-    }
-
-    public boolean isLeft() {
-        return left.isDefined();
-    }
-
-    public boolean isRight() {
-        return right.isDefined();
-    }
-
-    public ImEither<R,L> swap() {
-        return new ImEither<>(right, left);
-    }
-
-    public Option<R> toOption() {
-        return right;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.Optional;
+
+public interface ImEither<L,R> extends Serializable {
```

Comment:
Just a silly formatting note, I guess we normally leave a blank line between method definitions